daemon/src/daemon.rs

```diff
@@ -872,13 +877,45 @@ impl Daemon {
     }
 }
 
+// Spawn a file watcher and feed its events to the document_handle.
+// In addition, a short timeout after the last event, do a full re-scan, so that we don't miss any
+// file changes - the watcher isn't necessarily exhaustive.
 async fn spawn_file_watcher(base_dir: &Path, document_handle: DocumentActorHandle) {
     let mut watcher = Watcher::new(base_dir);
+
     tokio::spawn(async move {
-        while let Some(watcher_event) = watcher.next().await {
-            document_handle
-                .send_message(DocMessage::FromWatcher(watcher_event))
-                .await;
+        let debounce_duration = Duration::from_millis(100);
+
+        let debounce_timer = tokio::time::sleep(debounce_duration);
+        // Sleep does not implement the Unpin trait, so in order to use it with select!, we have to
+        // pin it first (according to the documentation).
+        tokio::pin!(debounce_timer);
+
+        let mut rescan_required = false;
+
+        loop {
+            tokio::select! {
+                maybe_event = watcher.next() => {
+                    if let Some(watcher_event) = maybe_event {
+                        document_handle
+                            .send_message(DocMessage::FromWatcher(watcher_event))
+                            .await;
+
+                        debounce_timer.as_mut().reset(tokio::time::Instant::now() + debounce_duration);
+                        rescan_required = true;
+                    } else {
+                        // Watcher terminated. Seems we're shutting down.
+                        return;
+                    }
+                }
+
+                _ = &mut debounce_timer, if rescan_required => {
```

Comment:
This looks like a confusing bit of syntax 😬 What I'd assume it does given the context of the PR is to catch the event that the `debounce_timer` ran out, guarded by `rescan_required`. So `debounce_timer` is a `Future` and this branch is only activated when it has an actual result. Am I right? :)

Reply:
I think I'm just not used to the syntax of `select!`... It seems I understood it correctly.

Btw, the [documentation](https://docs.rs/tokio/latest/tokio/time/struct.Sleep.html) uses a different construct:
```rust
() = &mut sleep => {
```
which seems to be equivalent. Both of these comments are just me processing the code while reading it, no action required :)